crates/rocksdb-store/src/chain_state/db.rs

```diff
@@ -21,78 +21,59 @@ impl ChainstateDb {
-        get_first_idx::<ChainstateSchema>(&self.db)
+        get_first_idx::<WriteBatchSchema>(&self.db)
     }
 
     fn get_last_idx(&self) -> DbResult<Option<u64>> {
-        get_last_idx::<ChainstateSchema>(&self.db)
+        get_last_idx::<WriteBatchSchema>(&self.db)
     }
 }
 
 impl ChainstateDatabase for ChainstateDb {
-    fn get_earliest_state_idx(&self) -> DbResult<u64> {
-        match self.get_first_idx()? {
-            Some(idx) => Ok(idx),
-            None => Err(DbError::NotBootstrapped),
-        }
-    }
-
-    fn get_last_state_idx(&self) -> DbResult<u64> {
-        match self.get_last_idx()? {
-            Some(idx) => Ok(idx),
-            None => Err(DbError::NotBootstrapped),
-        }
-    }
-
-    fn get_writes_at(&self, idx: u64) -> DbResult<Option<strata_state::state_op::WriteBatch>> {
-        Ok(self.db.get::<WriteBatchSchema>(&idx)?)
-    }
-
-    // TODO: define what toplevel means more clearly
-    fn get_toplevel_state(
-        &self,
-        idx: u64,
-    ) -> DbResult<Option<strata_state::chain_state::Chainstate>> {
-        Ok(self.db.get::<ChainstateSchema>(&idx)?)
-    }
-
-    fn write_genesis_state(
-        &self,
-        toplevel: &strata_state::chain_state::Chainstate,
-    ) -> DbResult<()> {
+    fn write_genesis_state(&self, toplevel: strata_state::chain_state::Chainstate) -> DbResult<()> {
         let genesis_key = 0;
+
+        // This should only ever be called once.
         if self.get_first_idx()?.is_some() || self.get_last_idx()?.is_some() {
             return Err(DbError::OverwriteStateUpdate(genesis_key));
         }
-        self.db.put::<ChainstateSchema>(&genesis_key, toplevel)?;
+
+        let fake_wb = WriteBatch::new_replace(toplevel);
+        self.db.put::<WriteBatchSchema>(&genesis_key, &fake_wb)?;
+
         Ok(())
     }
 
-    fn write_state_update(
-        &self,
-        idx: u64,
-        batch: &strata_state::state_op::WriteBatch,
-    ) -> DbResult<()> {
+    fn put_write_batch(&self, idx: u64, batch: strata_state::state_op::WriteBatch) -> DbResult<()> {
         if self.db.get::<WriteBatchSchema>(&idx)?.is_some() {
             return Err(DbError::OverwriteStateUpdate(idx));
         }
 
-        let pre_state_idx = idx - 1;
-        let pre_state = match self.db.get::<ChainstateSchema>(&pre_state_idx)? {
-            Some(state) => state,
-            None => return Err(DbError::OooInsert("Chainstate", idx)),
-        };
-        let post_state = state_op::apply_write_batch_to_chainstate(pre_state, batch);
+        // Make sure we always have a contiguous range of batches.
+        // FIXME this *could* be a race condition / TOCTOU issue, but we're only
+        // going to be writing from a single thread anyways so it should be fine
+        match self.get_last_idx()? {
+            Some(last_idx) => {
+                if idx != last_idx + 1 {
+                    return Err(DbError::OooInsert("Chainstate", idx));
```

Comment:
This assumes that there can't be updates to a given idx. I am not sure if that is valid, won't we want to update the entry in case of forks?

Reply:
Ah this is probably a mistake in translation.

Reply:
Oh wait no this was correct.  If we want to update an idx then we need to explicitly roll it back first.